src/activations.jl

```diff
@@ -256,32 +255,17 @@ about twice the error of Julia's, about twice the speed
     ifelse(x2 < 66f0, x * (n / d), sign(x))
 end
 
-# tanh_fast(x::Float16) # probably better. tanh_fast(x::Real) is very badly behaved for Float16.
-
-@inline function tanh_fast(x::Float16)
-    x2 = abs2(x)
-    n = evalpoly(x2, (Float16(1.0), Float16(0.1346), Float16(0.003597), Float16(2.235e-5)))
-    d = evalpoly(x2, (Float16(1.0), Float16(0.468), Float16(0.02626), Float16(0.0003455), Float16(9.0e-7)))
-    ifelse(x2 < Float16(18), x * (n / d), sign(x))
-end
-
-"""
-    tanh_fast(x) = exp(2x) / (1 + exp(2x))
-
-This is faster, but slightly less accurate, version of `tanh`.
-For `x::Float64`, this has errors up to about `3e-15`, sometimes `> 10 * eps(x)`,
-and should be about 2-5 times faster.
-
-But `tanh_new(x::Float32)` is more important, and a different approximation.
-"""
 @inline function tanh_fast(x::Real)
-   exp2x = @fastmath exp(x + x)
-   y = (exp2x - 1) / (exp2x + 1)
-   ifelse(x > 30, one(y), ifelse(x < -30, -one(y), y))
+    exp2x = @fastmath exp(x + x)
```

Comment:
What about
```julia
 @inline exp_inline(x) = Base.Math.exp_impl(x, Val(:ℯ))
```
and defining an associated diff rule, for the versions of Julia that define an inline `Base.Math.exp_impl`?

This should let the compiler SIMD the code.

Reply:
I saw this in the Zulip thread and had a go... but for me this is somewhere between the `@fastmath` variant and the plain `exp` in speed:
```
julia> @btime @. $ys = tanh_fast($xs);  # with @fastmath exp(x + x), and polynomial etc, M1 + Rosetta
  30.125 μs (0 allocations: 0 bytes)

julia> @btime @. $ys = tanh_fast($xs);  # with exp_inline
  47.208 μs (0 allocations: 0 bytes)

julia> @btime @. $ys = tanh_fast($xs);  # with Base.exp
  85.125 μs (0 allocations: 0 bytes)
```

Reply:
What about `Base.Math.exp_impl_fast(x, Val(:ℯ))`

Reply:
(I think that needs 1.7)

Reply:
Using `Base.Math.exp_impl_fast` seems to make no difference at all, strangely? I don't see an `@inline`.